src/core/google-auth.js

```diff
@@ -72,3 +74,251 @@ XMPPSocialProvider.prototype.login = function(loginOpts, continuation) {
+    function(refreshToken, email) {
+  this.storage.set('Google-Refresh-Token-Last',
+      JSON.stringify({refreshToken: refreshToken, email: email}));
+};
+
+XMPPSocialProvider.prototype.loadLastRefreshTokenAndEmail_ = function() {
+  return new Promise(function(fulfill, reject) {
+    this.storage.get('Google-Refresh-Token-Last').then(function(data) {
+      fulfill(JSON.parse(data));
+    }.bind(this));
+  }.bind(this));
+};
+
+XMPPSocialProvider.prototype.getEmail_ = function(accessToken) {
+  return new Promise(function(fulfill, reject) {
+    var xhr = freedom["core.xhr"]();
+    xhr.open('GET', 'https://www.googleapis.com/oauth2/v1/userinfo?alt=json', true);
+    xhr.on("onload", function() {
+      xhr.getResponseText().then(function(responseText) {
+        fulfill(JSON.parse(responseText).email);
+      }.bind(this));
+    }.bind(this));
+    xhr.setRequestHeader('Authorization', 'Bearer ' + accessToken);
+    xhr.send();
+  }.bind(this));
+};
+
+XMPPSocialProvider.prototype.getCode_ = function(
+    oauth, stateObj, forcePrompt, authUserEmail) {
+  var getCodeUrl = 'https://accounts.google.com/o/oauth2/auth?' +
+           'response_type=code' +
+           '&access_type=offline' +
+           '&client_id=' + this.oAuthClientId +
+           '&scope=' + this.oAuthScope +
+           (forcePrompt ? '&approval_prompt=force' : '') +
+           '&redirect_uri=' + encodeURIComponent(stateObj.redirect) +
+           '&state=' + encodeURIComponent(stateObj.state);
+
+  var url;
+  if (authUserEmail) {
+    // Skip account chooser and set the authuser param
+    url = getCodeUrl + '&authuser=' + authUserEmail;
+  } else {
+    // Got to account chooser.
+    url = 'https://accounts.google.com/accountchooser?continue=' +
+        encodeURIComponent(getCodeUrl);
+  }
+
+  return oauth.launchAuthFlow(url, stateObj).then(function(responseUrl) {
+    return responseUrl.match(/code=([^&]+)/)[1];
+  }.bind(this));
+};
+
+XMPPSocialProvider.prototype.tryToGetRefreshToken_ = function(
+    forcePrompt, authUserEmail) {
+  return new Promise(function(fulfill, reject) {
+    var oauth = freedom["core.oauth"]();
+    return oauth.initiateOAuth(this.oAuthRedirectUris).then(function(stateObj) {
+      this.getCode_(
+          oauth, stateObj, forcePrompt, authUserEmail).then(function(code) {
+        var data = 'code=' + code +
+            '&client_id=' + this.oAuthClientId +
+            '&client_secret=' + this.clientSecret +
+            "&redirect_uri=" + encodeURIComponent(stateObj.redirect) +
+            '&grant_type=authorization_code';
+        var xhr = freedom["core.xhr"]();
+        xhr.open('POST', 'https://www.googleapis.com/oauth2/v3/token', true);
+        xhr.setRequestHeader(
+            'content-type', 'application/x-www-form-urlencoded');
+        xhr.on('onload', function() {
+          xhr.getResponseText().then(function(responseText) {
+            fulfill(JSON.parse(responseText));
```

Comment:
suggestion: comment that this will have an access token but might not have a refresh token (my understanding)